design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,144 @@
+- Dynamically applying affinity rules to VMIs
+
+## Definition of Users
+A user with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+## User Stories
+As a VM user, I would like to dynamically increase or decrease the number of CPUs used by a VM without requiring the VM to restart.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+In VM objects, the existing CPU `sockets` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
+Under the `cpu` entry, it will be possible to define a maximum number of sockets. That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set.
+
+Increasing the number of CPU sockets will not only add CPUs to the guest but also increase the CPU resources available to the virt-launcher pod, via a migration (more on that in the Implementation section).
+The way LibVirt handles CPU hotplug is by exposing the maximum number of CPUs to the guest and turning off the unused CPUs. In this document, when we refer to "adding" CPUs to the guest, we're really just turning them on. This is why we need to define a maximum number of sockets in advance.
+
+## Drawbacks / Limitations
+- Since CPU hot-(un)plug involves a live migration, VMs will have to be `LiveMigratable` to enable CPU hotplug
+- Each disabled vCPU, so `(maxSockets - sockets) * cores * thread`, consumes 8MiB of overhead memory
+- This feature will be incompatible with CPU requests/limits, at least initially, since those values need to dynamically scale according to the current number of enabled CPUs
+
+## API Examples
+
+### Spec API
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+  name: vm-cirros
+spec:
+  running: false
+  liveUpdateFeatures:
+  - cpu
+      maximumSockets: 4
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 4
+          threads: 2
+```
+In this example, the user will be able to double the guest CPU resources by increasing the value of `sockets`.
+
+### Status API
+
+The VMI status must indicate the current and desired number of CPU sockets. These values need to be propagated back to the VM status to allows the user to have an indication of when their changes applied to the running VMI.
+
+The user should be able to inspect the VM.Status values below to determine the state of their changes. This is similar in concept to someone inspecting a Deployments status to determine the desired number of pods vs the number of pods that are available.
+
+```
+// The declared number of sockets the vm is converging on
+vm.Status.Resources.Sockets.Desired
+
+// The number of sockets the vm has currently active.
+vm.Status.Resources.Sockets.Ready
```

Comment:
you might find that having this desired and ready sockets info is useful on the vmi.Status as well. if that' sthe case, it can originate on the vmi.status and be back propagated to the vm for visibility. I think that's similar to how the deployment and replicaset controllers work.